v2/yarpcrouter/router.go

```diff
@@ -58,19 +58,43 @@ type MapRouter struct {
 func NewMapRouter(defaultService string) MapRouter {
 	return MapRouter{
 		defaultService:            defaultService,
-		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.Procedure),
+		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.TransportProcedure),
 		serviceNames:              map[string]struct{}{defaultService: {}},
 	}
 }
 
+// NewMapRouterWithEncodingProcedures constructs a new MapRouter with the given default service name and registers
+// the given encoding-level procedures as transport-level procedures
+func NewMapRouterWithEncodingProcedures(defaultService string, encodingProcedures []yarpc.EncodingProcedure) MapRouter {
```

Comment:
Let’s open this to discussion. I think we should choose between a Register API or a constructor API.

We could keep the router construct as simple as `New(name string, []TransportProcedure)`, staying low-level and expose a function to build []TransportProcedure from the transport and encoding middleware chains and []EncodingProcedure. Also, we still want to be able to register transport procedures that do not have an encoding layer! Transport-level procedure definitions can do things like encoding agnostic shadowing, fan-out, etc.

In general, I favor layered APIs and making the user experience easy at the Fx layer.

Reply:
We could. What do you think @kriskowal?

Reply:
Sounds good to me, let's make the Fx layer easy. I like the idea of creating a `TransportProcedure` from the middleware chains since it would simplifies many concerns.

My `ClientProvider` changes in #1594 intentionally mimicked the router so whichever we choose we should be consistent. I'm a bit biased since I just landed my changes, so I like the `Register` method :p. 

However, not having a `Register` function signals immutability, which I think makes this the most desirable API.

Reply:
🤔 Call me crazy, but, @kriskowal, we should also consider the case where a procedure may want to rely on the Router for a probably terrible reason. Enabling users to register post object creation could resolve some circular dependencies.

..Or maybe this is an even better reason to go with the constructor solution haha

Reply:
Sounds like a better reason to go with the constructor and immutable object solution.

Reply:
haha yeah agreed. I'll put up a corresponding change to the `yarpcclient.Provider` to drop the `Register` method